File: Recursion/quick_sort.py
arr = [1, 6, 9, 7, 0, 2, 3, 8]


def quick_sort(arr, pivot):
    temp = arr[:pivot]
    temp2 = arr[pivot:len(arr)]
    quick_sort(temp, len(temp)//2)
    quick_sort(temp2, len(temp2)//2)

File: sorting/quick_sort.c
#include <stdio.h>

int partition(int arr[], int low, int high) {
    int pivot = arr[high]; // Choose last element as pivot
    int i = low - 1;

    for (int j = low; j < high; j++) {
        if (arr[j] <= pivot) {
            i++;
            // swap arr[i] and arr[j]
            int temp = arr[i]; arr[i] = arr[j]; arr[j] = temp;
        }
    }
    // swap arr[i+1] and arr[high] (pivot)
    int temp = arr[i + 1]; arr[i + 1] = arr[high]; arr[high] = temp;
    return i + 1;
}

void quick_sort(int arr[], int low, int high) {
    if (low < high) {
        int pi = partition(arr, low, high);
        quick_sort(arr, low, pi - 1);
        quick_sort(arr, pi + 1, high);
    }
}
